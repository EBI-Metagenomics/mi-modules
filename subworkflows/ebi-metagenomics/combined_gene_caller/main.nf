
include { PRODIGAL                 } from '../../../modules/ebi-metagenomics/prodigal/main'
include { FRAGGENESCAN             } from '../../../modules/ebi-metagenomics/fraggenescan/main'
include { COMBINEDGENECALLER_MERGE } from '../../../modules/ebi-metagenomics/combinedgenecaller/merge/main'

workflow COMBINED_GENE_CALLER {

    take:
    // Should the mask_file be merged into the assembly channel ?
    ch_assembly  // channel: [ val(meta), [ fasta ] ]
    ch_mask_file // channel: [ val(meta), [.out | .txt] ]

    main:

    ch_versions = Channel.empty()

    PRODIGAL ( ch_assembly, channel.value("sco") )
    ch_versions = ch_versions.mix(PRODIGAL.out.versions.first())

    FRAGGENESCAN ( ch_assembly )
    ch_versions = ch_versions.mix(FRAGGENESCAN.out.versions.first())

    ch_mask_file = ch_mask_file ?: Channel.empty()

    ch_annotations = PRODIGAL.out.gene_annotations.join(
        PRODIGAL.out.nucleotide_fasta, by: [0]
    ).join(
        PRODIGAL.out.amino_acid_fasta, by: [0]
    ).join(
        FRAGGENESCAN.out.gene_annotations, by: [0]
    ).join(
        FRAGGENESCAN.out.nucleotide_fasta, by: [0]
    ).join(
        FRAGGENESCAN.out.amino_acid_fasta, by: [0]
    ).join(
        ch_mask_file, by: [0], remainder: true
    )

    COMBINEDGENECALLER_MERGE ( ch_annotations )

    ch_versions = ch_versions.mix(COMBINEDGENECALLER_MERGE.out.versions.first())

    emit:
    faa      = COMBINEDGENECALLER_MERGE.out.faa  // channel: [ val(meta), [ faa ] ]
    ffn      = COMBINEDGENECALLER_MERGE.out.ffn  // channel: [ val(meta), [ ffn ] ]
    versions = ch_versions                       // channel: [ versions.yml ]
}

