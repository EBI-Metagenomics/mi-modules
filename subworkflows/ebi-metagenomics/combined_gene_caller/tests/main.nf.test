nextflow_workflow {

    name "Test Subworkflow COMBINED_GENE_CALLER"
    script "../main.nf"
    workflow "COMBINED_GENE_CALLER"
    config "./nextflow.config"

    tag "subworkflows"
    tag "subworkflows_ebimetagenomics"
    tag "subworkflows/combined_gene_caller"
    tag "combinedgenecaller/merge"
    tag "prodigal"
    tag "fraggenescan"

    test("it should work properly with a fasta as input") {

        when {
            workflow {
                """
                input[0] = [
                    [ id:'test', single_end:false ],
                    file("${moduleDir}/tests/fixtures/ERZ19591644_FASTA_subsample.fasta.gz", checkIfExists: true)
                ]
                input[1] = channel.empty()
            """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
