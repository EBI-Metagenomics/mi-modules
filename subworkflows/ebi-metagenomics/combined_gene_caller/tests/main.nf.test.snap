{
    "it should work properly with a fasta as input": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.faa.gz:md5,d8354102b645c1b1b97e3aef559b8a64"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.ffn.gz:md5,9b7a1ce6c2380e7be072e68c171d93cd"
                    ]
                ],
                "2": [
                    "versions.yml:md5,14ce365346593b2444c8be16843e3444",
                    "versions.yml:md5,6ce9415d5682dd2a5a1ad52ef05477a3",
                    "versions.yml:md5,e19d0b42602a46bc22cca0bdc9c71a49"
                ],
                "faa": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.faa.gz:md5,d8354102b645c1b1b97e3aef559b8a64"
                    ]
                ],
                "ffn": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.ffn.gz:md5,9b7a1ce6c2380e7be072e68c171d93cd"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,14ce365346593b2444c8be16843e3444",
                    "versions.yml:md5,6ce9415d5682dd2a5a1ad52ef05477a3",
                    "versions.yml:md5,e19d0b42602a46bc22cca0bdc9c71a49"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.2"
        },
        "timestamp": "2025-01-15T16:47:31.100271"
    }
}
